activestorage/lib/active_storage/service/gcs_service.rb

```diff
@@ -160,7 +176,37 @@ def bucket
       end
 
       def client
-        @client ||= Google::Cloud::Storage.new(**config.except(:bucket, :cache_control))
+        @client ||= Google::Cloud::Storage.new(**config.except(:bucket, :cache_control, :iam, :gsa_email))
+      end
+
+      def issuer
+        @issuer ||= if @config[:gsa_email]
+          @config[:gsa_email]
+        else
+          uri = URI.parse("http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/email")
+          http = Net::HTTP.new(uri.host, uri.port)
+          request = Net::HTTP::Get.new(uri.request_uri)
+          request["Metadata-Flavor"] = "Google"
+
+          http.request(request).body
```

Comment:
What should happen if this request doesn't return a successful response?

Reply:
If it fails an error will be raised that says the metadata server can't be contacted, should I rescue it and return a custom error?

Reply:
I added error handling here, https://github.com/rails/rails/pull/42723/commits/dfd19e1a76bd313bdfb35dc02c2e53f71611b953. I can squash if this approach looks good, wasn't sure what errors to raise so I added two new gcs specific error classes.